compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PolymorphicSignatureLowering.kt

```diff
@@ -0,0 +1,87 @@
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.ir.copyTypeParametersFrom
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.config.LanguageFeature
+import org.jetbrains.kotlin.ir.builders.declarations.addValueParameter
+import org.jetbrains.kotlin.ir.builders.declarations.buildFun
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.isNullableAny
+import org.jetbrains.kotlin.ir.util.dump
+import org.jetbrains.kotlin.ir.util.hasAnnotation
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.name.FqName
+
+internal val polymorphicSignaturePhase = makeIrFilePhase(
+    ::PolymorphicSignatureLowering,
+    name = "PolymorphicSignature",
+    description = "Replace polymorphic methods with fake ones according to types at the call site"
+)
+
+private val polymorphicSignatureAnnotation = FqName("java.lang.invoke.MethodHandle.PolymorphicSignature")
+
+class PolymorphicSignatureLowering(val context: JvmBackendContext) : IrElementTransformerVoid(), FileLoweringPass {
+    override fun lower(irFile: IrFile) {
+        if (context.state.languageVersionSettings.supportsFeature(LanguageFeature.PolymorphicSignature))
+            irFile.transformChildrenVoid()
+    }
+
+    // If the return type is Any?, then it is also polymorphic (e.g. MethodHandle.invokeExact
+    // has polymorphic return type, while VarHandle.compareAndSet does not).
+    override fun visitTypeOperator(expression: IrTypeOperatorCall): IrExpression =
+        (expression.argument as? IrCall)?.takeIf { it.type.isNullableAny() }?.transform(expression.typeOperand)
+            ?: super.visitTypeOperator(expression)
+
+    override fun visitCall(expression: IrCall): IrExpression =
+        expression.transform(null) ?: super.visitCall(expression)
+
+    private fun IrCall.transform(castedReturnType: IrType?): IrCall? {
+        val function = symbol.owner as? IrSimpleFunction ?: return null
+        if (!function.hasAnnotation(polymorphicSignatureAnnotation))
+            return null
+        assert(function.valueParameters.singleOrNull()?.varargElementType != null) {
+            "@PolymorphicSignature methods should only have a single vararg argument: ${dump()}"
+        }
+
+        val values = (getValueArgument(0) as IrVararg?)?.elements?.map {
+            when (it) {
+                is IrExpression -> it
+                is IrSpreadElement -> it.expression // `*xs` acts as `xs` (for compatibility?)
```

Comment:
> (for compatibility?)

More like there's no other option here. We don't know what's inside the array under the spread, so we can't specialize the signature. But maybe spread operator should be disallowed on polymorphic method calls, I suppose we can think about it later.

Reply:
>But maybe spread operator should be disallowed on polymorphic method calls

That's what I initially expected to happen. The behavior of spreads is extremely weird right now, for example a spread and positional arguments after it are not concatenated into a single array.

Reply:
> a spread and positional arguments after it are not concatenated into a single array

:o

Please report an [issue](https://kotl.in/issue), so that we don't forget to address it before 1.4.

Reply:
Created [KT-35226](https://youtrack.jetbrains.com/issue/KT-35226)